python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -785,59 +687,7 @@ def stream(
                 def my_dbt_assets(context, dbt: DbtCliResource):
                     yield from dbt.cli(["run"], context=context).stream()
         """
-        has_any_parallel_tasks = self.should_fetch_row_count
-
-        if not has_any_parallel_tasks:
-            # If we're not enqueuing any parallel tasks, we can just stream the events in
-            # the main thread.
-            yield from self._stream_asset_events()
-            return
-
-        if self.should_fetch_row_count:
-            logger.info(
-                "Row counts will be fetched for non-view models once they are materialized."
-            )
-
-        # We keep a list of emitted Dagster events and pending futures which augment
-        # emitted events with additional metadata. This ensures we can yield events in the order
-        # they are emitted by dbt.
-        output_events_and_futures: List[Union[Future, DbtDagsterEventType]] = []
-
-        # Point at project directory to ensure dbt adapters run correctly
-        with pushd(str(self.project_dir)), ThreadPoolExecutor(
-            max_workers=STREAM_EVENTS_THREADPOOL_SIZE
-        ) as executor:
-            self._dbt_run_thread = executor.submit(
-                self._stream_dbt_events_and_enqueue_postprocessing,
-                output_events_and_futures,
-                executor,
-            )
-
-            # Step through the list of output events and futures, yielding them in order
-            # once they are ready to be emitted
-            event_to_emit_idx = 0
-            while True:
-                all_work_complete = get_future_completion_state_or_err(
-                    [self._dbt_run_thread, *output_events_and_futures]
-                )
-                if all_work_complete and event_to_emit_idx >= len(output_events_and_futures):
-                    break
-
-                if event_to_emit_idx < len(output_events_and_futures):
-                    event_to_emit = output_events_and_futures[event_to_emit_idx]
-
-                    if isinstance(event_to_emit, Future):
-                        # If the next event to emit is a Future (waiting on postprocessing),
-                        # we need to wait for it to complete before yielding the event.
-                        try:
-                            event = event_to_emit.result(timeout=0.1)
-                            yield event
-                            event_to_emit_idx += 1
-                        except:
-                            pass
-                    else:
-                        yield event_to_emit
-                        event_to_emit_idx += 1
+        return DbtEventIterator(self._stream_asset_events(), self)
```

Comment:
Looks like this is the remaining call site for `._stream_asset_events`. Could we just move its implementation in the constructor of `DbtEventIterator`?

```suggestion
        return DbtEventIterator(self)
```

Reply:
Hmm, trying to wrap my head around this. I think we need `DbtEventIterator` to accept any iterator as input (e.g. when we return a `DbtEventIterator` from `fetch_row_counts()`), so we wouldn't want to call `_stream_asset_events` in that case.

We could do like `DbtEventIterator.from_cli_invocation(self)` which would contain the body of `_stream_asset_events()`, that seems decent